New draft name: draft-ietf-ipsecme-ikev2-multiple-ke-00

smyslov opened the pull request:
Changes:
1. New draft name as a result of WG adoption and generalization of an approach
2. New dedicated exchange type IKE_FOLLOWUP_KE
3. Get rid of nonces in additional key exchanges
4. Abstract is rewritten (added explicit indication of what is changed in RFC7296, it seems that this is now a requirement if some RFC is being changed)
5. A lot of clarifications in the text

Commit:
New draft name: draft-ietf-ipsecme-ikev2-multiple-ke-00

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+	<abstract>
+        <t> This document describes how to extend the Internet Key Exchange Protocol
+		Version 2 (IKEv2) to allow multiple key exchanges to take place 
+        while computation of a shared secret during Security Association (SA) setup.
```
```suggestion
        while computing of a shared secret during a Security Association (SA) setup.
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+        post-quantum key exchanges in conjunction with the classical (Elliptic Curve) Diffie-Hellman key exchange,
+        so that the resulting shared key is resistant against quantum computer attacks.
+        Another possible application is a possibility to combine several key exchanges 
+        in situations when no single key exchange algorithm is trusted by both parties.
```
```suggestion
        in situations when no single key exchange algorithm is trusted by both initiator and responder.
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+
+		<section title="Proposed Extension" >
+            <t>
+			This document describes a framework to perform multiple successive key 
```
```suggestion
			This document describes a method to perform multiple successive key 
```

ctjhai replied:
Given that we are dropping the work `framework` from the title, I thought it is not necessary to use this word in the body. This could be just me, but the work `framework` implies something of a high-level or theoretical, whereas in this draft, pretty details description is given on how to do multiple key-exchanges. However, I'm happy to use this word if others agree to leave it in.

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+		<section title="Proposed Extension" >
+            <t>
+			This document describes a framework to perform multiple successive key 
+            excahnges in IKEv2. This framework allows to integrate QSC in IKEv2, while
```
```suggestion
            exchanges in IKEv2. It allows integration of QSC in IKEv2, while
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+			This document describes a framework to perform multiple successive key 
+            excahnges in IKEv2. This framework allows to integrate QSC in IKEv2, while
+			maintaining backwards compatibility, to derive a set of IKE keys that
+			have resistance to quantum computer attacks.  The framework allows the
```
```suggestion
			is resistant to quantum computer attacks.  This extension allows the
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+			such that should the post-quantum secrets not be present, the derived
+			shared secret is equivalent to that of the standard IKEv2; on the
+			other hand, a post-quantum shared secret is obtained if both classical
+			and post-quantum key exchange data are present.  This framework also
```
```suggestion
			and post-quantum key exchange data are present.  This extension also
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+			mechanism, we do an initial key exchange, using a smaller, possibly
+			non-quantum resistant primitive, such as ECDH.  Then, before we do
+			the IKE_AUTH exchange, we perform one or more IKE_INTERMEDIATE exchanges,
+			each of which includes an additional key exchange.  As the IKE_INTERMEDIATE
```
```suggestion
			each of which contains an additional key exchange.  As the IKE_INTERMEDIATE
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+
+            <t> Note also, that the proposed approach of performing multiple successive key exchanges
+            in such a way that resulting session keys depend on all of them is not limited
+            to achieving quantum resistance only. In can also be used when all 
```
```suggestion
            to achieving quantum resistance only. It can also be used when all 
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+            The remainder of this document is organized as follows.  <xref target="design"/>
+            summarizes design criteria.  <xref target="framework"/> describes how
+    	    multiple key exchanges are performed between two IKE peers and how
+    	    keying materials are derived for both SAs and child SAs.  A summary of alternative
```
```suggestion
    	    keying materials are derived for both SAs and Child SAs.  A summary of alternative
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+
+	<section title="Design Criteria" anchor="design">
+        <t>
+        The design of the proposed framework is driven by the
```
```suggestion
        The design of the proposed extension is driven by the
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+    
+    	<t><list style="hanging" hangIndent="5"><t hangText="1)">
+    	Need for post-quantum cryptography in IPsec.  Quantum computers
+    	might become feasible in the next 5-10 years.  If current
```
Don't feel like we should speculate on this time frame, so how about the following?

```suggestion
    	might become feasible in the near future.  If current
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+    	Hybrid.  Currently, there does not exist a post-quantum key
+    	exchange that is trusted at the level that ECDH is trusted
+        against conventional (non-quantum) adversaries.  A hybrid
+        approach allows introducing promising post-quantum candidates
```
I think the use of the word candidates implies that we only need hybrid while NIST is standardising PQC. However, it is quite likely that hybrid may still be used even after standardisation. As we stated in some parts of this documents, some of the PQC algorithms are relative new and are not that well-trusted yet. So how about this change:

```suggestion
        post-quantum algorithms to be introduced
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+    	or "non post-quantum" and does not create assumptions
+        about the properties of the algorithms, meaning that if
+        algorithms with different properties become necessary in the future,
+        this framework can be used unchanged to facilitate migration to
```
```suggestion
        this extension can be used unchanged to facilitate migration to
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+    
+    	<t hangText="8)">
+    	Deterministic operation.  This requirement means that the hybrid
+    	post-quantum exchange, and thus, the computed key, will be based
```
`keys`?

```suggestion
    	post-quantum exchange, and thus, the computed keys, will be based
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+    	</t>
+    
+    	<t hangText="11)">
+    	FIPS compliance.  IPsec is widely used in Federal Information
```
```suggestion
    	Federal Information Processing Standards (FIPS) compliance.  IPsec is widely used in Federal Information
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+
+	</section>
+
+	<section title="The Framework of Hybrid Post-Quantum Key Exchange" anchor="framework">
```
```suggestion
	<section title="Multiple Key Exchanges" anchor="framework">
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+			and post-quantum key exchanges, nor post-quantum algorithms which are
+			true key exchanges versus post-quantum algorithms that act as key
+			transport mechanisms; all are treated equivalently by the
+			protocol. To make this more clear for implementers this document
```
```suggestion
			protocol. To be more specific, this document
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+<-- {IKE_AUTH} -->
+            ]]></artwork></figure>
+
+            <t> The additional key exchanges can use algorithms that are
```
```suggestion
            <t> The additional key exchanges may use algorithms that are
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+            <t> The additional key exchanges can use algorithms that are
+    		currently considered to be resistant to quantum computer attacks.  These
+    		algorithms are collectively referred to as post-quantum algorithms in
+    		this document.</t>
```
Perhaps we should add the following sentence to emphasise that classical primitives could be used too:

```suggestion
    		this document. However, it is also possible to use classical (EC)DH primitives for non post-quantum requirements.</t>
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+                </t>
+
+                <t> If the IKE SA was created using multiple key exchange methods, the peers 
+                may want continue using multiple key exchanges in the CREATE_CHILD_SA exchange too.
```
```suggestion
                may want to continue using multiple key exchanges in the CREATE_CHILD_SA exchange too.
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+                          <--  HDR(CREATE_CHILD_SA), SK {SA, Nr, KEr,
+                                   N(ADDITIONAL_KEY_EXCHANGE)(link1)}
+
+HDR(IKE_FOLLOWUP_KE), SK {KEi1,
```
I thought we should write it following your symbols, i.e. KEi(n) and KEr(n)?

```suggestion
HDR(IKE_FOLLOWUP_KE), SK {KEi(1),
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+
+HDR(IKE_FOLLOWUP_KE), SK {KEi1,
+ N(ADDITIONAL_KEY_EXCHANGE)(link1)} -->
+                                 <--  HDR(IKE_FOLLOWUP_KE), SK {KEr1,
```
```suggestion
                                 <--  HDR(IKE_FOLLOWUP_KE), SK {KEr(1),
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+                                 <--  HDR(IKE_FOLLOWUP_KE), SK {KEr1,
+                                   N(ADDITIONAL_KEY_EXCHANGE)(link2)}
+
+HDR(IKE_FOLLOWUP_KE), SK {KEi2,
```
```suggestion
HDR(IKE_FOLLOWUP_KE), SK {KEi(2),
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+
+HDR(IKE_FOLLOWUP_KE), SK {KEi2,
+ N(ADDITIONAL_KEY_EXCHANGE)(link2)} -->
+                                 <--  HDR(IKE_FOLLOWUP_KE), SK {KEr2,
```
```suggestion
                                 <--  HDR(IKE_FOLLOWUP_KE), SK {KEr(2),
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+                                 <--  HDR(IKE_FOLLOWUP_KE), SK {KEr2,
+                                   N(ADDITIONAL_KEY_EXCHANGE)(link3)}
+
+HDR(IKE_FOLLOWUP_KE), SK {KEi3,
```
```suggestion
HDR(IKE_FOLLOWUP_KE), SK {KEi(3),
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+
+HDR(IKE_FOLLOWUP_KE), SK {KEi3,
+ N(ADDITIONAL_KEY_EXCHANGE)(link3)} -->
+                                 <--  HDR(IKE_FOLLOWUP_KE), SK {KEr3}
```
```suggestion
                                 <--  HDR(IKE_FOLLOWUP_KE), SK {KEr(3)}
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+                notification and the content of this notify doesn't correspond to any active key exchange
+                state the responder has, it MUST send back a new error type notification
+                STATE_NOT_FOUND. This is a non-fatal error notification, its Notify Message Type is &lt;TBA by IANA&gt;, 
+                Protocol ID and SPI Size are both set to 0 and the data is empty. If initiator receives
```
```suggestion
                Protocol ID and SPI Size are both set to 0 and the data is empty. If the initiator receives
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+                start a new CREATE_CHILD_SA exchange (eventually followed by the IKE_FOLLOWUP_KE exchanges)
+                to retry the failed attempt. If the initiator continues to receive
+                STATE_NOT_FOUND notifications after several retries, it MUST treat this situation 
+                as fatal error and delete IKE SA (sending DELETE payload).
```
```suggestion
                as fatal error and delete IKE SA by sending a DELETE payload.
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+                as fatal error and delete IKE SA (sending DELETE payload).
+                </t>
+
+                <t> When rekeying IKE SA or Child SA it is possible that the peers start doing this
```
```suggestion
                <t> When rekeying IKE SA or Child SA, it is possible that the peers start doing this
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+                </t>
+
+                <t> In most cases rekey collisions are resolved in the CREATE_CHILD_SA exchange.
+                However, a situation may occur when due to packet loss one of the peers receives CREATE_CHILD_SA message
```
```suggestion
                However, a situation may occur when due to packet loss, one of the peers receives CREATE_CHILD_SA message
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+                IKE_FOLLOWUP_KE exchange with next key exchange.
+                </t>
+
+                <t> In most cases rekey collisions are resolved in the CREATE_CHILD_SA exchange.
```
```suggestion
                <t> In most cases, rekey collisions are resolved in the CREATE_CHILD_SA exchange.
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+                However, a situation may occur when due to packet loss one of the peers receives CREATE_CHILD_SA message
+                requesting rekeying SA that is already being rekeyed by this peer (i.e. the CREATE_CHILD_SA
+                exchange initiated by this peer has been already completed and the series of IKE_FOLLOWUP_KE exchanges is in progress).
+                In this case TEMPORARY_FAILURE notification MUST be sent in response to such request.
```
```suggestion
                In this case, a TEMPORARY_FAILURE notification MUST be sent in response to such request.
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+SKEYSEED = prf(SK_d, KE  | Ni | Nr | KE(1) | ... KE(n))
+                ]]></artwork></figure>
+
+                <t> In case of Child SA creating or rekey:
```
`creation?`

```suggestion
                <t> In case of Child SA creation or rekey:
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+KEYMAT = prf+ (SK_d, KE  | Ni | Nr | KE(1) |  ... KE(n))
+                ]]></artwork></figure>
+
+                <t> In both cases SK_d is from existing IKE SA; KE, Ni, Nr - shared key and nonces
```
```suggestion
                <t> In both cases SK_d is from existing IKE SA; KE, Ni, Nr are the shared key and nonces
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+                ]]></artwork></figure>
+
+                <t> In both cases SK_d is from existing IKE SA; KE, Ni, Nr - shared key and nonces
+                from the CREATE_CHILD_SA; KE(1)..KE(n) - shared keys from additional key exchanges.
```
```suggestion
                from the CREATE_CHILD_SA respectively; KE(1)...KE(n) are the shared keys from additional key exchanges.
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+    	sufficient resistance to  quantum attacks.  Accordingly the
+    	post-quantum security level achieved is at least 128 bits.</t>
+    
+    	<t>SKEYSEED is calculated from shared, KEx, using an algorithm defined
```
```suggestion
    	<t>SKEYSEED is calculated from shared KE(x) using an algorithm defined
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+    
+    	<t>SKEYSEED is calculated from shared, KEx, using an algorithm defined
+    	in Transform Type 2.  While a quantum attacker may learn the value
+    	of KEx', if this value is obtained by means of a classical key exchange,
```
```suggestion
    	of KE(x), if this value is obtained by means of a classical key exchange,
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+    	<t>SKEYSEED is calculated from shared, KEx, using an algorithm defined
+    	in Transform Type 2.  While a quantum attacker may learn the value
+    	of KEx', if this value is obtained by means of a classical key exchange,
+    	other KEx values generated by means of a quantum-resistant algorithm
```
```suggestion
    	other KE(x) values generated by means of a quantum-resistant algorithm
```

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```suggestion
        the time of the connection, for example by mounting a man-in-the-middle (MitM)
```

ctjhai reviewed (approved):
Happy with the draft and there are some minor suggestions.

ctjhai reviewed (approved):
Happy with the draft and there are some minor suggestions.

ctjhai commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1037 @@
+    	</t>
+    
+        <t hangText="12)">
+        Ability to use this framework with multiple classical (EC)DH key exchanges.
```
```suggestion
        Ability to use this method with multiple classical (EC)DH key exchanges.
```

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

I don't mind, but then we should get rid of word "framework" throughout the document, shouldn't we?

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

smyslov committed:
Apply suggestions from code review

Co-Authored-By: ctjhai <ctjhai@users.noreply.github.com>

danvangeest commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1036 @@
+            <xref target="RFC7383"/> is available.</t>
+
+            <t> The fact that all Additional Key Exchange Transform Types share
+            the same registry with Transform Type 4 allows additional key exchnages
```
exchnages -> exchanges

smyslov replied:
Yes, typo.

danvangeest commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1036 @@
+        <t> This document describes how to extend the Internet Key Exchange Protocol
+		Version 2 (IKEv2) to allow multiple key exchanges to take place 
+        while computing of a shared secret during a Security Association (SA) setup.
+        The primary application of this feature in IKEv2 is an ability to perform one or more 
```
"an ability" -> "the ability"

smyslov replied:
I agree.

danvangeest commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1036 @@
+        The primary application of this feature in IKEv2 is an ability to perform one or more 
+        post-quantum key exchanges in conjunction with the classical (Elliptic Curve) Diffie-Hellman key exchange,
+        so that the resulting shared key is resistant against quantum computer attacks.
+        Another possible application is a possibility to combine several key exchanges 
```
"a possibility" -> "the ability" (funny to have "possibl.*" twice so close in the same sentence).

smyslov replied:
Yes.

danvangeest commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1036 @@
+			too.  IKE does have a mechanism to handle fragmentation within
+			UDP <xref target="RFC7383"/>, however that is only applicable to
+			messages exchanged after the IKE_SA_INIT.  To use this mechanism,
+			this specification rely on the IKE_INTERMEDIATE exchange as outlined in
```
"rely" -> "relies"

smyslov replied:
Yes.

danvangeest commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1036 @@
+            </t>
+
+            <t> In order to support IKE fragmentation for additional key exchanges
+            that may have a long public keys, the proposed framework utilizes the IKE_INTERMEDIATE exchange 
```
"a long public keys" -> "long public keys"

smyslov replied:
My bad :-)

danvangeest commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1036 @@
+            the same registry with Transform Type 4 allows additional key exchnages
+            to be of any type - either post-quantum ones or classical (EC)DH ones.
+            This approach allows any combination of defined key exchange methods
+            to take place. This also allows to perform a single post-quantum 
```
"allows to perform" -> "allows performing"

smyslov replied:
OK.

danvangeest commented on draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:
```diff
@@ -0,0 +1,1036 @@
+                (along with Transform Type 4) and the responder agrees to perform additional
+                key exchanges, then the additional key exchanges are performed in a series 
+                of new IKE_FOLLOWUP_KE exchanges that follows the CREATE_CHILD_SA exchange.
+                The IKE_FOLLOWUP_KE exchange is introduced as a dedicated exchange type to transfer data of additional key exchanges,
```
"additional key exchanges, that follow" -> "additional key exchanges following"

smyslov replied:
OK.

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

smyslov committed:
Update draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
